mypy/main.py

```diff
@@ -62,6 +94,20 @@ def main(script_path: Optional[str], args: Optional[List[str]] = None) -> None:
         args = sys.argv[1:]
     sources, options = process_options(args)
 
+    # Set Python version if given Python executable, but the version is default
```

Comment:
If `python-version` trumps `python-executable`, one would need to always pass `python-version` if you passed `python-executable` (unless you are on the default 3.6), which defeats the purpose of this inference AIUI. Unless `options.python_version` defaults to `None`, in which case these will be isomorphisms I think. Do you want `python-version` to silently set `python-executable` even if both are given? I'm not sure I understand what you mean by "trumps". If they disagree I consider it an error and exit.

Reply:
> If they disagree I consider it an error and exit.

That's not accurate in the corner case of `--python-version=3.6 --python-executable=python27` because of how you handle the default. You're on the right track with having `options.python_version` default to `None`, which will resolve that case.

Reply:
Hm, this is hard.

We could make it so that if you specify either, it provides the default for the other, and if you specify both, they must correspond. For this to work the default of `options.python_version` must be changed to `None` so that you can reliably check whether it was specified or not. (If we didn't do that, you wouldn't be able to tell the difference between `--python-version 3.6` and omitting it.)

But now what should we do when you specify neither? The current situation (without this PR) is that it defaults to `3.6` (i.e. `defaults.PYTHON3_VERSION`) regardless of what you use to run mypy with. But then if you use some Python version other than 3.6 to run mypy, it wouldn't have a sensible default for `--python-executable`.

So this leads us to consider using the Python version and executable used to run mypy as the default for both -- but this means changing the current default, which is always 3.6.

But hey! **The docs are wrong.** They state (in several places) that the default for `--python-version`is "whatever version of Python is running mypy", in direct conflict with what is actually implemented. What's strangest is that I vividly recall that it used to do it that way and we changed it, but I can't find evidence that we ever did it that way in the logs. The default is currently 3.6, before that it was 3.5, and before that there was only the `--py2` flag to force Python 2, otherwise it was Python 3, and there was nothing to force a specific minor version (there was also nothing that depended on it). Even when the `--python-version=<major>.<minor>` option was introduced (in b3bc75cb66cc0f965ccdadf58b2247f551f7f6a6) I can't find any use of the minor version. (The docs for this were introduced in 897398eff895969af05b7d573e9ad72b972fc3d4 and they were wrong then, and nobody noticed -- I reviewed it...)

So now I've talked myself into believing that it's fine to do what the docs said we were doing in the first place, even though we never did it that way. I've filed https://github.com/python/mypy/issues/4620 to ask for other opinions just in case. But for now I think you should go ahead and implement it the way you'd like it -- just change the default in `options.py` to `None` so we can tell the difference between true absence and an explicit version equal to the default.

Note that this still assumes that in most cases the user should specify `--python-version`, not `--python-executable`, since we can usually find the appropriate Python executable. The new default just makes it more likely the people who don't care that much and have a custom Python installation will not be inconvenienced as long as they don't specify either of these options.

Reply:
>We could make it so that if you specify either, it provides the default for the other, and if you specify both, they must correspond.

I think we all agree this is the ideal state.

> just change the default in options.py to None so we can tell the difference between true absence and an explicit version equal to the default.

"Just". Sadly there is a lot of code that assumes that the version is set, but I will work around this as best I can :)

Also +1 on doing what the docs say.

Reply:
> Sadly there is a lot of code that assumes that the version is set

I think you might be on an adjacent page here. The `Options` constructor will set it to `None`, but the options parser will replace these `None`s with `sys.executable` and `sys.version_info[:2]`

Reply:
I was hesitant to modify the option parsing due to the coherence check in `testargs.py`. I think I am okay with changing it just to check attribute names at this point.

Reply:
Also I should add, even if it is set at runtime mypy doesn't seem to want to narrow the type :(

Reply:
Moving `python_version` and `python_executable` to `special-opts` seems like the best approach here

Reply:
Yes I actually was just experimenting with the same idea 😄